Assets/Scenes/Scripts/GameManager.cs

```diff
@@ -15,34 +15,51 @@ public class GameManager : MonoBehaviour {
     void PlayerManager() {
         player.transform.Translate(new Vector3(Input.GetAxis("Horizontal"), Input.GetAxis("Vertical"), 0) * Time.deltaTime * playerMoveSpeed);
         projectileSpawnTime += Time.deltaTime;
-        if (projectileSpawnTime >= 1f / playerMainClassLevel) { // level determines how fast the player can shoot, TODO: extract this for multiple classes
+        if (projectileSpawnTime >= 1f / playerMainClassLevel) {
             GameObject projectile = Instantiate(projectilePrefab, player.transform.position + new Vector3(1, 0, 0), Quaternion.identity);
             projectiles.Add(projectile);
             StartCoroutine(SpiralAndFade(projectile));
             projectileSpawnTime = 0;
         }
+        if (player.GetComponent<Collider2D>().IsTouchingLayers(LayerMask.GetMask("Enemy"))) {
+            playerHealth -= enemyAttackDamage; // Handle player damage (Example : play sound, show effect)
+        }
     }
     private IEnumerator SpiralAndFade(GameObject obj) { // Spriral and fade effect for projectiles #1 animation for projectiles
         float angle = 0, startTime = Time.time;
         Renderer objRenderer = obj.GetComponent<Renderer>();
         Color originalColor = objRenderer.material.color;
-        while (Time.time - startTime < fadeDuration) {
+        while (Time.time - startTime < fadeDuration && obj != null) {
             angle += spiralSpeed * Time.deltaTime;
             obj.transform.position = player.transform.position + new Vector3(Mathf.Cos(angle), Mathf.Sin(angle), 0) * Mathf.Min(Time.time - startTime, 2.0f);
             objRenderer.material.color = new Color(originalColor.r, originalColor.g, originalColor.b, Mathf.Lerp(1, 0, (Time.time - startTime) / fadeDuration));
             yield return null;
         }
-        Destroy(obj);
+        if (obj != null) {
+            Destroy(obj);
+        }
     }
     void EnemyManager() {
         enemySpawnTime += Time.deltaTime;
         if (enemySpawnTime >= enemySpawnInterval) {
             enemies.Add(Instantiate(enemyPrefab, player.transform.position + (Vector3)Random.insideUnitCircle.normalized * 10f, Quaternion.identity));
             enemySpawnTime = 0;
         }
-        foreach (var enemy in enemies) {
-            if (enemy != null) {
+        List<GameObject> enemiesCopy = new List<GameObject>(enemies);
+        foreach (var enemy in enemiesCopy) {
```

Comment:
**enemies** are already a List of Game Object that contains multiple enemy, does it really need to re-declare?

like in **line 52** that take each projectile from projectiles that contains multiple projectile like enemy.

Reply:
so This don't that much necessary because it used to avoid modifying the list while iterating over it like removing enemies but for the projectiles list you don't modify the list during iteration so you don't need to create a copy

Reply:
got it!